src/trie_tree.rs

```diff
@@ -0,0 +1,475 @@
+        if self.is_empty() {
+            return Ok(V::zero());
+        }
+        Ok(self.store.get_leaf(key)?.unwrap_or_else(V::zero))
+    }
+
+    /// Generate merkle proof
+    pub fn merkle_proof(&self, mut keys: Vec<H256>) -> Result<MerkleProof> {
+        if keys.is_empty() {
+            return Err(Error::EmptyKeys);
+        }
+
+        // sort keys
+        keys.sort_unstable();
+
+        // Collect leaf bitmaps
+        let mut leaves_bitmap: Vec<H256> = Default::default();
+        for current_key in &keys {
+            let mut bitmap = H256::zero();
+            for height in (0..=core::u8::MAX).rev() {
+                let parent_key = current_key.parent_path(height);
+                let parent_branch_key = BranchKey::new(height, parent_key);
+                if let Some(parent_branch) = self.store.get_branch(&parent_branch_key)? {
+                    let (sibling, target) = if current_key.is_right(height) {
+                        (parent_branch.left, parent_branch.right)
+                    } else {
+                        (parent_branch.right, parent_branch.left)
+                    };
+
+                    match target {
+                        MergeValue::ShortCut { key, .. } => {
+                            if !sibling.is_zero() {
+                                bitmap.set_bit(height);
+                            }
+                            if !key.eq(current_key) {
+                                let fork_height = key.fork_height(current_key);
+
+                                bitmap.set_bit(fork_height);
+                                break;
+                            }
+                        }
+                        _ => {
+                            if !sibling.is_zero() {
+                                bitmap.set_bit(height);
+                            }
+                        }
+                    }
+                }
+            }
+            leaves_bitmap.push(bitmap);
+        }
+
+        let mut proof: Vec<MergeValue> = Default::default();
+        let mut stack_fork_height = [0u8; MAX_STACK_SIZE]; // store fork height
+        let mut stack_top = 0;
+        let mut leaf_index = 0;
+
+        while leaf_index < keys.len() {
+            let leaf_key = keys[leaf_index];
+            let fork_height = if leaf_index + 1 < keys.len() {
+                leaf_key.fork_height(&keys[leaf_index + 1])
+            } else {
+                core::u8::MAX
+            };
+
+            let mut skip_height = Vec::new();
+            for height in 0..=fork_height {
+                if stack_top > 0 && stack_fork_height[stack_top - 1] == height {
+                    stack_top -= 1;
+                    skip_height.push(height);
+                }
+                if stack_top == 0 {
+                    break;
+                }
+            }
+
+            let mut proof_result = Vec::new();
+            for height in (0..=fork_height).rev() {
```

Comment:
suggest to use the filtered heights directly to reduce the `vec#contains` check

```suggestion
             let heights = (0..=fork_height).into_iter().filter(|height|
                if stack_top > 0 && stack_fork_height[stack_top - 1] == *height {
                    stack_top -= 1;
                    false
                } else {
                    true
                }
            ).collect::<Vec<_>>();

            let mut proof_result = Vec::new();
            for height in heights.iter().copied().rev() {
```